[Fix #12049] Add new `Style/HashFetchChain` cop to detect method chains that can be replaced with a single call to `dig`

dvandersluis opened the pull request:
```ruby
# bad
hash.try(:[], 'foo').try(:[], 'bar')
hash.fetch('foo', nil)&.fetch('bar', nil)

# good
hash.dig('foo', 'bar')
```

Fixes #12049.

-----------------

Before submitting the PR make sure the following are checked:

* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Ran `bundle exec rake default`. It executes all tests and runs RuboCop on its own code.
* [x] Added an entry (file) to the [changelog folder](https://github.com/rubocop/rubocop/blob/master/changelog/) named `{change_type}_{change_description}.md` if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.

[1]: https://chris.beams.io/posts/git-commit/

dvandersluis commented on lib/rubocop/cop/style/hash_dig.rb:
```diff
@@ -0,0 +1,115 @@
+      #
+      # @example
+      #   # bad
+      #   hash.try(:[], 'foo').try(:[], 'bar')
```
Fair point about activesupport, I forgot it wasn't part of ruby itself. I'm ok with removing the checks for `try`, too. It might make sense to have it when called with safe navigation, but that kind of removes the purpose of calling `try` in the first place...

Earlopain replied:
I agree with this, thanks (:

Earlopain commented on lib/rubocop/cop/style/hash_dig.rb:
```diff
@@ -0,0 +1,105 @@
+      #   hash.dig('foo', 'bar')
+      #
+      #   # good - not handled by the cop since the final `fetch` value is non-nil
+      #   hash.fetch('foo', {}).fetch('bar', {})
```
I looked through some code [here](https://github.com/search?q=%2F%5C.fetch%5C%28%5B%5E%2C%5C%29%5D*%2C+%28nil%7C%5C%7B%5C%7D%29%5C%29%26%3F%5C.fetch%2F&type=code) and I think this can also be handled:

```rb
stream.fetch('tags', {}).fetch('title', '')
# into
stream.dig('tags', 'title') || ''
```

Does that make sense or can it mess something up? In some cases it might get complicated with precedence.

dvandersluis replied:
I think we can do this but it’s less straightforward as to what the right correction is because `dig(…).fetch('foo', default)` could conceivably be just as good (when the chain is length 3+). 

With limiting to nil final default values it’s simple and can be fully replaced by a single `dig`.

Earlopain replied:
Right, makes sense. Thanks for explaining.

koic commented on lib/rubocop/cop/style/hash_dig.rb:
```diff
@@ -0,0 +1,105 @@
+        def last_fetch_non_nil?(node)
+          # When chaining `fetch` methods, `fetch(x, {})` is acceptable within
+          # the chain, as long as the last method in the chain has a `nil`
+          # default value.
+
+          return false unless node.method?(:fetch)
```
If the blank line has no meaning, it can be removed. Alternatively, this comment could be used as a method comment.

```suggestion
          # default value.
          return false unless node.method?(:fetch)
```

koic commented:
As noted in the documentation, `Style/DigChain` is mentioned, so I'm concerned that a more specific name like `Lint/HashFetchNilDefaultChain` would be better than `Style/HashDig`. `Style/HashDig` is a very generic name, but the rule it actually enforces seems quite limited.

koic commented:
Since `Hash#dig` was introduced in Ruby 2.3, it would be better to specify `minimum_target_ruby_version: 2.3`. This prevents issues that may arise when replacing `Hash#fetch` with `Hash#dig`.

koic commented:
Consideration could be given to adding this rule to the style guide:
https://github.com/rubocop/ruby-style-guide

dvandersluis commented:
Agreed, `HashDig` is too generic given that the cop has been reduced to only considering chains of `fetch`. I renamed it to `Style/HashFetchChain` which is slightly terser than the suggestion, and inline with `Style/DigChain`. I've also added `minimum_target_ruby_version`.

chaadow commented:
@dvandersluis hey, this breaks when fetch has no arguments. i'm using a gem that has a `fetch` method which has no relation to hashes

Example:
```ruby
call = client.api.calls(external_id.to_s).fetch
```

here is the error : `/Users/chedli/.rbenv/versions/3.4.2/lib/ruby/gems/3.4.0/bundler/gems/rubocop-d7be4961df6a/lib/rubocop/cop/style/hash_fetch_chain.rb:82:in 'RuboCop::Cop::Style::HashFetchChain#last_fetch_non_nil?'`

chaadow commented:
@dvandersluis it's the twilio gem here is an example : https://github.com/twilio/twilio-ruby?tab=readme-ov-file#fetch-a-single-sms-message-by-sid

koic commented:
@chaadow I've confirmed that error and opened #https://github.com/rubocop/rubocop/pull/14022 to address it.

chaadow commented:
@koic Thanks!

jasonkarns commented:
I feel it should also be noted that this cop is not safe if the hash (or any intermediate hash) hash has a default value or default_proc configured.

`[]` and `dig` will return the default value (or run the default proc), whereas `fetch` will raise key-error. So replacing chained `fetch` with `dig` can result in differing behavior.

Should default-values and/or default_procs be called out in the docs for this? That seems like an obscure edge case that is exactly the kind of thing that would go unnoticed when (perhaps blindly) accepting rubocop suggestions.

jasonkarns commented on lib/rubocop/cop/style/hash_fetch_chain.rb:
```diff
@@ -0,0 +1,105 @@
+      #
+      # @safety
+      #   This cop is unsafe because it cannot be guaranteed that the receiver
+      #   is a `Hash` or that `fetch` or `dig` have the expected standard implementation.
```
Should probably also mention here that it's further unsafe because the hash (or intermediate hashes) could have `default` and/or `default_proc` configured, which would be respected for `dig` but not fetch.
